### PR TITLE
refactor(utils): move output-target related utils to @utils

### PR DESCRIPTION
**Pull request by alicewriteswrongs**
This moves a bunch of utility functions (and some constants) from src/compiler/output-targets/output-utils.ts to instead live in a new file in the `@utils` directory (`src/utils`). This is done to reflect the fact that functions in the file are used outside of the `src/compiler/` bundle so they should be pulled out into a common location.

## Pull request checklist

Please check if your PR fulfills the following requirements:
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [ ] Build (`npm run build`) was run locally and any changes were pushed
- [ ] Unit tests (`npm test`) were run locally and passed
- [ ] E2E Tests (`npm run test.karma.prod`) were run locally and passed
- [ ] Prettier (`npm run prettier`) was run locally and passed

## Pull request type

 

 

Please check the type of change your PR introduces:
- [ ] Bugfix
- [ ] Feature
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

GitHub Issue Number: N/A

## What is the new behavior?

-
-
-

## Does this introduce a breaking change?

- [ ] Yes
- [ ] No

### CODE REVIEW

File: src/cli/task-docs.ts [hunk 1]
```diff
@@ -1,4 +1,5 @@
-import { isOutputTargetDocs } from '../compiler/output-targets/output-utils';
```
Comment (alicewriteswrongs) [inline src/cli/task-docs.ts:1]: when I was fiddling with build stuff I noticed this relative import from `src/compiler/` which was causing code from within the compiler bundle be also pulled into the `cli` bundle. I considered a bit how to address it - we could export the output targets related utils (and constants) from the compiler bundle, making this into something like

```ts
import { isOutputTargetDocs } from '@stencil/core'
```

but I thought that wasn't too great for a few reasons, mainly that it changes the public API of `@stencil/core`. I also thought just for organization, if the output target related utils are used across the codebase they should really live in the `src/utils` directory and therefore be importable from the `@utils` namespace. So I thought we should look at moving the functions to `src/utils/output-target.ts` and moving the constants to `src/utils/constants.ts`

File: scripts/bundles/internal-platform-hydrate.ts [hunk 2]
```diff
@@ -39,6 +39,9 @@ export async function internalHydrate(opts: BuildOptions) {
       banner: getBanner(opts, 'Stencil Hydrate Platform'),
       preferConst: true,
     },
+    treeshake: {
+      moduleSideEffects: false,
+    },
```
Comment (alicewriteswrongs) [inline scripts/bundles/internal-platform-hydrate.ts:44]: the `internal-platform-hydrate` bundle depends on functions which use the output-target utils which, in turn, imports `path`. However, none of the functions which it uses actually _use_ functioned exported by `path`, so when bundling it doesn't actually import anything from `path` but if you don't set this option Rollup will leave an unqualified import like so:

```ts
import "path";
```

if you say `moduleSideEffects: false` you're basically saying 'dont assume that imported a module has a desired side effect' which allows an import that doesn't result in any symbols being pulled into the importing module to be eliminated.

Another way to say this is that without this change the treeshaking validation for `test.dist` doesn't pass for this bundle.

**Review by alicewriteswrongs — COMMENTED** [review]
just noting a few things

File: scripts/bundles/internal-platform-hydrate.ts (hunk 2 above)
Comment (rwaskiewicz) [reply scripts/bundles/internal-platform-hydrate.ts:44]: Can we capture this info in a comment for future us?

**Review by rwaskiewicz — COMMENTED** [review]
Had one minor comment/ask for JSDoc

File: src/compiler/config/outputs/validate-dist.ts [hunk 3]
```diff
@@ -11,7 +8,10 @@ import {
   DIST_TYPES,
   getComponentsDtsTypesFilePath,
   isOutputTargetDist,
-} from '../../output-targets/output-utils';
+} from '@utils';
```
Comment (tanner-reits) [inline src/compiler/config/outputs/validate-dist.ts:11]: Can we combine this with the existing `@utils` import?

File: src/compiler/config/outputs/validate-www.ts [hunk 4]
```diff
@@ -1,15 +1,8 @@
 import { buildError, isBoolean, isString } from '@utils';
+import { COPY, DIST_GLOBAL_STYLES, DIST_LAZY, isOutputTargetDist, isOutputTargetWww, WWW } from '@utils';
```
Comment (tanner-reits) [inline src/compiler/config/outputs/validate-www.ts:2]: Can we combine this with the existing @utils import?

File: src/compiler/fs-watch/fs-watch-rebuild.ts [hunk 5]
```diff
@@ -1,8 +1,8 @@
 import { isString, unique } from '@utils';
+import { isOutputTargetDocsJson, isOutputTargetDocsVscode, isOutputTargetStats } from '@utils';
```
Comment (tanner-reits) [inline src/compiler/fs-watch/fs-watch-rebuild.ts:2]: Can we combine this with the existing @utils import?

**Review by tanner-reits — CHANGES_REQUESTED** [review]
Couple small things (sorry for the repetition). Also agree w/ Ryan's comment about adding docs/comment for the `moduleSideEffects` flag.

File: src/compiler/config/outputs/validate-dist.ts (hunk 3 above)
Comment (alicewriteswrongs) [reply src/compiler/config/outputs/validate-dist.ts:11]: oops, thought I captured all of those!

**Comment by alicewriteswrongs** [comment]
just added a comment explaining the `moduleSideEffects` option for Rollup and then also bundled up some imports

**Comment by alicewriteswrongs** [comment]
@tanner-reits you inspired me: https://github.com/ionic-team/stencil/pull/4233

**Review by tanner-reits — APPROVED** [review]
Looks good!

**Review by rwaskiewicz — APPROVED** [review]

**Pull request merged**